File: content/showcase/benchmarks/runtime/hashmap/zig/hashmap.zig
const std = @import("std");

const Rng = struct {
    state: u64,
    fn next(self: *Rng) u64 {
        var x = self.state;
        x ^= x << 13;
        x ^= x >> 7;
        x ^= x << 17;
        self.state = x;
        return x;
    }
};

pub fn main() !void {
    const n: u64 = 5_000_000;
    var m = std.AutoHashMap(u64, u64).init(std.heap.page_allocator);
    defer m.deinit();
    var keys = Rng{ .state = 0x9E3779B97F4A7C15 };
    var i: u64 = 0;
    while (i < n) : (i += 1) try m.put(keys.next(), i);
    var hits = Rng{ .state = 0x9E3779B97F4A7C15 };
    var misses = Rng{ .state = 0x2545F4914F6CDD1D };
    var found: u64 = 0;
    var sum: u64 = 0;
    i = 0;
    while (i < n) : (i += 1) {
        const hit_key = hits.next();
        const miss_key = misses.next();
        const key = if (i % 2 == 0) hit_key else miss_key;
        if (m.get(key)) |v| {
            found += 1;
            sum += v;
        }
    }
    try std.io.getStdOut().writer().print("inserted: {d} found: {d} sum: {d}\n", .{ m.count(), found, sum });
}

File: content/showcase/benchmarks/runtime/hashmap/rust/hashmap.rs
use std::collections::HashMap;

struct Rng(u64);

impl Rng {
    fn next(&mut self) -> u64 {
        let mut x = self.0;
        x ^= x << 13;
        x ^= x >> 7;
        x ^= x << 17;
        self.0 = x;
        x
    }
}

fn main() {
    let n: u64 = 5_000_000;
    let mut m: HashMap<u64, u64> = HashMap::new();
    let mut keys = Rng(0x9E3779B97F4A7C15);
    for i in 0..n {
        m.insert(keys.next(), i);
    }
    let mut hits = Rng(0x9E3779B97F4A7C15);
    let mut misses = Rng(0x2545F4914F6CDD1D);
    let mut found = 0u64;
    let mut sum = 0u64;
    for i in 0..n {
        let hit_key = hits.next();
        let miss_key = misses.next();
        let key = if i % 2 == 0 { hit_key } else { miss_key };
        if let Some(&v) = m.get(&key) {
            found += 1;
            sum += v;
        }
    }
    println!("inserted: {} found: {} sum: {}", m.len(), found, sum);
}
